src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -4578,6 +4618,33 @@ BoundExpression bindElement(CollectionElementSyntax syntax, BindingDiagnosticBag
             }
         }
 
+        private TypeSymbol? InferCollectionLiteralType(ImmutableArray<BoundExpression> initializers)
+        {
+            var useSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded; // PROTOTYPE: Include use-site diagnostics.
+            var bestType = BestTypeInferrer.InferBestType(initializers, Conversions, ref useSiteInfo, out _);
+
+            if (bestType is { } && !bestType.IsVoidType())
+            {
+                // Check that each element (even those without a type) can be converted to the best type.
+                foreach (var initializer in initializers)
+                {
+                    if (!Conversions.ClassifyImplicitConversionFromExpression(initializer, bestType, ref useSiteInfo).Exists)
```

Comment:
So it's possible to choose a best type for a set of expressions, even if one of the expressions is not convertible to that type? That seems so strange.

Reply:
Why would we clear this? It seems like the better experience would be to let it be found as the best type and then have conversion errors?

Reply:
> So it's possible to choose a best type for a set of expressions, even if one of the expressions is not convertible to that type?

`BestTypeInferrer.InferBestType()` ignores expressions that do not contribute a type, so we'll choose a type of `int` for `[1, null]` for instance.

> Why would we clear this? It seems like the better experience would be to let it be found as the best type and then have conversion errors?

Perhaps, although for reference, this matches the behavior for `switch` expressions, see `SwitchExpressionBinder.InferResultType()`.